include/pybind11/factory_support.h

```diff
@@ -0,0 +1,280 @@
+/*
+    pybind11/factory_support.h: init factory function implementation and support code.
+
+    Copyright (c) 2017 Jason Rhinelander <jason@imaginary.ca>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#pragma once
+
+#include "class_support.h"
+
+NAMESPACE_BEGIN(pybind11)
+NAMESPACE_BEGIN(detail)
+NAMESPACE_BEGIN(factory)
+
+// Proxy object to accept an instance with conversion to a value_and_holder, but with a custom type
+// caster that does type checking (so that you can't pass invalid types into a `__init__`).
+template <typename T> struct v_h_proxy { value_and_holder v_h; };
+
+NAMESPACE_END(factory)
+
+// Simple, load-only type caster that does a type check and value_and_holder construction.
+template <typename T> class type_caster<factory::v_h_proxy<T>> {
+private:
+    const type_info *tinfo = get_type_info(typeid(T));
+    factory::v_h_proxy<T> value;
+public:
+    bool load(handle src, bool /* convert */) {
+        if (!src || !tinfo) return false;
+        auto *inst = reinterpret_cast<instance *>(src.ptr());
+        value.v_h = inst->get_value_and_holder(tinfo, false);
+        return value.v_h.inst != nullptr;
+    }
+    operator factory::v_h_proxy<T> &() { return value; }
+    template <typename> using cast_op_type = factory::v_h_proxy<T> &;
+    static PYBIND11_DESCR name() { return type_descr(_<T>()); }
+};
+
+NAMESPACE_BEGIN(factory)
+
+inline void no_nullptr(void *ptr) {
+    if (!ptr) throw type_error("pybind11::init(): factory function returned nullptr");
+}
+
+// Makes sure the `value` for the given value_and_holder is preallocated; if it isn't, preallocate
+// it.  Returns the value pointer reference.
+inline void *&preallocate(value_and_holder &v_h) {
+    if (!v_h) v_h.value_ptr() = v_h.type->operator_new(v_h.type->type_size);
+    return v_h.value_ptr();
+}
+
+// Makes sure the `value` for the given value_and_holder is not allocated; if it is, deallocate it.
+// Returns the (null) value pointer reference.
+inline void *&deallocate(value_and_holder &v_h) {
```

Comment:
One possible case is when you have overloaded constructors when you have mixed placement new and init factory constructors, such as:

```
cl.def("__init__", [](Class &self, double i) { new (&self) Class(i); });
cl.def("__init__", py::init([](int j) { return new Class(j); });
```
when called from Python as `Class(3)` to go to the second overload.  Because the lazy allocation happens during argument loading, the `&self` will get preallocated even though the overload fails, and the other init factories need to deal with it.

I suppose we could deprecated placement new constructors and get rid of this eventually (i.e. in an eventual 3.0), in which case we wouldn't need to worry about this case.

Re: comments, that's a fair criticism; I'll edit them to be more "why" than "what".